### ISSUE TITLE
Does yolov8-pose use information about edges between keypoints in the MS COCO keypoint dataset format ?

### ISSUE THREAD

**Issue by medphisiker** [ultralytics/ultralytics#8500]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, thank you for a great framework and model )

Question: Does yolov8-pose use information about edges between keypoints in the MS COCO keypoint dataset format ?

I read documentation about yolov8-pose dataset format([link](https://docs.ultralytics.com/ru/datasets/pose/)).
P.S. I tried to switch the site documentation to English, but it didn't work out)

As far as I understand it, the dataset in YOLO's Ultra lyrics format does not include information about edges between key points.
I also try to export two MS COCO keypoint datasets to YOLO pose format.
I create the first MS COCO key point dataset with edges between keypoints by CVAT.
This dataset has MS COCO keypoint skeleton section related to edges between keypoints:

```
    "categories": [
        {
            "id": 1,
            "name": "treska_bbox",
            "supercategory": "",
            "keypoints": [],
            "skeleton": []
        },
        {
            "id": 2,
            "name": "treska",
            "supercategory": "",
            "keypoints": [
                "eye",
                "cutt",
                "belly"
            ],
            "skeleton": [
                [
                    2,
                    3
                ],
                [
                    1,
                    2
                ]
            ]
        }
 ```
 
 Then I create the second MS COCO keypoint dataset with keypoints without edges between them.
 This dataset has not MS COCO keypoint skeleton section related to edges between keypoints:
 ````
     "categories": [
        {
            "id": 1,
            "name": "treska_bbox",
            "supercategory": "",
            "keypoints": [],
            "skeleton": []
        },
        {
            "id": 2,
            "name": "treska",
            "supercategory": "",
            "keypoints": [
                "eye",
                "cut",
                "belly"
            ],
            "skeleton": []
        }
````

Then I run this python script to convert both of them to YOLO pose format:
```
from ultralytics.data.converter import convert_coco


if __name__ == "__main__":
    labels_dir = "/path_to_ms_coco_keypoints_dataset_folder/annotations"

    convert_coco(labels_dir=labels_dir, use_segments=True)
```

And I get the identical result of converting these datasets with one picture.
Here is an `coco_converted/labels/person_keypoints_default/frame_000030.txt` file
```
0 0.479127 0.715719 0.301723 0.204354
```

I also wanted to ask if there are models that use information about edges between points to try to restore the positions of points that could not be found?
I have read this issue ([link](https://github.com/ultralytics/ultralytics/issues/8175)). 
As I understand it, usually models do not do this and an algorithmic approach is used? I was thinking maybe there is some kind of additive in the form of graph neural networks that will try to predict the positions of the undiscovered points of the skeleton?

**Comment by glenn-jocher** [comment]
@medphisiker hello! Thanks for reaching out and for your kind words about our framework. 😊

Regarding your question about YOLOv8-pose and the use of edges between keypoints in the MS COCO keypoint dataset format: YOLOv8-pose primarily focuses on detecting keypoints without explicitly utilizing the information about edges (skeleton) between these keypoints during the detection phase. The model is designed to identify and localize keypoints based on the visual features present in the image.

For the restoration of positions of points that might not be found directly, YOLOv8-pose does not inherently use edges or graph neural networks to predict missing keypoints. However, post-processing techniques, including algorithmic approaches or additional models, could potentially be employed to infer the positions of undetected keypoints based on the detected ones and the known skeletal structure typical for the objects of interest.

Your observation about the conversion process resulting in identical outputs for datasets with and without edge information is consistent with the current design, as the conversion script focuses on keypoints.

For advanced scenarios like predicting the positions of undiscovered points, exploring graph neural networks or other post-processing methods could be an interesting direction. While not directly supported in YOLOv8-pose, such approaches can complement the keypoints detected by the model to achieve more comprehensive pose estimation results.

I hope this clarifies your queries! If you have more questions or need further assistance, feel free to ask.

**Comment by medphisiker** [comment]
Thank you very much for the detailed answer)
```
The model is designed to identify and localize keypoints based on the visual features present in the image.
```
I also thought about this, however, considering the design of a more classical architecture of a two-stage detector Keypoints RCNN ([link](https://learnopencv.com/human-pose-estimation-using-keypoint-rcnn-in-pytorch/#overview)).

Unfortunately, I don't have an accurate imagination of the YOLOv8 architecture. I assumed that yolov8-pose is also looking for keypoints as masks which active pixels corresponding to one type of key point. I assumed that yolov8-pose is like yolov8-seg only with special masks that look for key points. 

I thought so, by drawing an analogy to the Keypoint RCNN.
I don't know if this hypothesis is correct =)

**Comment by glenn-jocher** [comment]
@medphisiker, glad you found the information helpful! 😊 Your curiosity about the architecture and approach is spot-on. YOLOv8, including its pose estimation variant, indeed takes a unique path compared to traditional two-stage detectors like Keypoint R-CNN.

While YOLOv8-pose doesn't explicitly use masks for keypoints in the same way YOLOv8-seg does for segmentation, it's an interesting analogy! YOLOv8-pose directly predicts keypoints from the image features learned during training, focusing on efficiency and speed while maintaining accuracy. It doesn't segment out keypoints but rather locates them as specific points within the image.

Your hypothesis draws a creative parallel between segmentation masks and keypoint detection. While not exactly how YOLOv8-pose operates, thinking about keypoints as specialized "masks" is a fascinating concept! YOLOv8's architecture is streamlined for direct keypoint prediction, making it distinct from the approach of Keypoint R-CNN, which indeed segments keypoints in a two-stage process.

Keep the great questions coming! If you're diving deeper into pose estimation or have more thoughts, we're here to chat. 😄

**Comment by medphisiker** [comment]
Thank you for the intresting and detailed answer)

```
YOLOv8-pose directly predicts keypoints from the image features learned during training, focusing on efficiency and speed while maintaining accuracy. 
```

Then I wanted to draw another analogy for finding key points )
We can use some backbone(feature extractor) form deep learning model for image classification(resnet, etc.) with fully connected layers to predict for example two coordinates(x,y) for human nose keypoint on human face image ) And we have a simple model for solving regression task on images for one keypoint.

Is this analogy closer to YOLOv8 way?)

**Comment by glenn-jocher** [comment]
@medphisiker, you're on the right track with your analogy! 😊 Indeed, YOLOv8-pose, much like the scenario you described, leverages a powerful backbone (akin to ResNet and others used in image classification) to extract rich features from images. On top of this feature extraction phase, it employs specialized layers designed to predict the coordinates (x, y) of various keypoints directly.

Your analogy of using a deep learning model's backbone for feature extraction, followed by regression to predict keypoints, aligns well with the essence of YOLOv8-pose's approach. It simplifies the process by integrating keypoint detection into the direct output of the network, optimizing for both speed and accuracy without the need for additional stages or segmentation tasks.

Here's a simplified conceptual snippet to illustrate:

```python
# Pseudo-code for conceptual understanding
features = backbone(image)  # Extract features using a backbone like ResNet
keypoints = regression_head(features)  # Predict keypoints directly from features
```

This streamlined process allows YOLOv8-pose to efficiently predict multiple keypoints across various objects within a single forward pass of the network. Great thinking! 🌟

**Comment by medphisiker** [comment]
Thank you for the details and the interesting story Glenn 😊
It was great to learn more about how yolov8-pose works 🚀

**Comment by glenn-jocher** [comment]
@medphisiker you're welcome! I'm glad you found the information helpful and interesting. If you have any more questions or need further clarification as you explore YOLOv8-pose, feel free to reach out. Happy coding! 😊🚀

**Comment by medphisiker** [comment]
I will definitely write when new questions arise. Thank you very much again) 😊🚀

**Comment by glenn-jocher** [comment]
@medphisiker you're welcome! 😊 Feel free to reach out anytime. Happy to help and looking forward to your questions. Happy coding! 🚀
